src/colvarcomp.h

```diff
@@ -1799,6 +1804,50 @@ class colvar::neuralNetwork
     virtual void apply_force(colvarvalue const &force);
 };
 
+#ifdef TORCH
+// only when LibTorch is available
+class colvar::torchANN
+  : public colvar::linearCombination
+{
+protected:
+    torch::jit::script::Module nn;
+    /// the index of nn output component
+    size_t m_output_index;
+    bool use_double_input;
+    //bool use_gpu;
+    // 1d tensor, concatenation of values of sub-cvcs
+    torch::Tensor input_tensor;
+    torch::Tensor nn_outputs;
+    torch::Tensor input_grad;
+    // record the initial index of of sub-cvcs in input_tensor
+    std::vector<int> cvc_indices;
+public:
+    torchANN(std::string const &conf);
+    virtual ~torchANN();
+    virtual void calc_value();
+    virtual void calc_gradients();
+    virtual void apply_force(colvarvalue const &force);
+
+    /// Redefined to handle periodicity
+    virtual cvm::real dist2(colvarvalue const &x1,
+			    colvarvalue const &x2) const;
+    /// Redefined to handle periodicity
+    virtual colvarvalue dist2_lgrad(colvarvalue const &x1,
+				    colvarvalue const &x2) const;
+    /// Redefined to handle periodicity
```

Comment:
Is it necessary to overload these functions to handle periodicity?

Reply:
It seems to me that it is necessary. I actually overloaded them some time ago in order to resolve an issue I had when running eABF with this class.   

These functions are called when `extendedLagrangian` is on and `f_cv_homogeneous=true`:
https://github.com/Colvars/colvars/blob/6ac3a18895b94b80b06b4ec68988092f63cb2a9d/src/colvar.cpp#L2230-L2234
Without overloading, potential energy and gradients in `colvar::update_extended_Lagrangian()`  are computed by calling the functions in the base class (periodicity is not considered), e.g. for the gradient,
https://github.com/Colvars/colvars/blob/6ac3a18895b94b80b06b4ec68988092f63cb2a9d/src/colvarcomp.cpp#L666-L670

In my case with eABF,  the issue turned out to be that due to the wrong gradient the extended variable $x_{ext}$ can not be pulled to cross the boundary after $x=\xi(X_t)$ is already wrapped to the other side of the boundary.

Reply:
I'm noticing here that only a single shift is performed when implementing the nearest-image convention. Given how general this type of variable can be, maybe simply copying the code from `distanceZ` would suffice? 
https://github.com/Colvars/colvars/blob/5e5c1d59021c48fcb25850e02c0f93d641b38f6c/src/colvarcomp_distances.cpp#L320-L329

I'll look into a way to just reuse functions like those in a derived class without resorting to the preprocessor macros.

Reply:
Right, it is safer to handle the general case.  For the moment I just copy the code from distanceZ class.